File: 2019/day-2/1202_program_alarm.rb
def get_input
  return IO.read('input.txt').split(',').map{ |i| i.to_i }
end

# For #reasons, before starting we must
# replace input_array[1] with 12 and input_array[2] with 2
def restore_gravity_assist_program(input_array)
  input_array[1] = 12
  input_array[2] = 2
  return input_array
end

# takes in an array of integers
# outputs the result of all the conversions we're instructed to do
def start_this_ridiculous_process(gravity_program)
  (0...gravity_program.length - 1).step(4).each do |i|
    value1 = gravity_program[gravity_program[i+1]]
    value2 = gravity_program[gravity_program[i+2]]
    output_position = gravity_program[i+3]
    # opcodes
    case gravity_program[i]
    when 1
      gravity_program[output_position] = value1 + value2
    when 2
      gravity_program[output_position] = value1 * value2
    when 99
      puts "The answer is #{gravity_program[0]}"
      return
    else
      puts "Oops this opcode is unexpected"
    end
  end
end

def solve_part_1
  input_array = get_input
  starting_input = restore_gravity_assist_program(input_array)
  start_this_ridiculous_process(starting_input)
end

solve_part_1

File: 2019/day-2/1202_program_alarm.js
const fs = require('fs');

const input = fs.readFileSync("input.txt").toString();

function adjustInputForProgram(input) {
  input_array = input.split(',').map(Number);
  input_array[1] = 12;
  input_array[2] = 2;
  return input_array;
}

function runGravityProgram(programArray) {
  let valueOnePosition, valueTwoPosition, outputPosition;
  for (let i = 0; i < programArray.length - 1; i += 4) {
    valueOnePosition = programArray[i+1];
    valueTwoPosition = programArray[i+2];
    outputPosition = programArray[i+3];
    switch(programArray[i]) {
      case 1:
        programArray[outputPosition] = programArray[valueOnePosition] + programArray[valueTwoPosition];
        break;
      case 2:
        programArray[outputPosition] = programArray[valueOnePosition] * programArray[valueTwoPosition];
        break;
      case 99:
        console.log(`The answer is ${programArray[0]}`);
        return;
      default:
        console.log("ruh roh, something's wrong");
    }
  }
}

runGravityProgram(adjustInputForProgram(input));